### PR TITLE
Refactor and enhance issue indexer to support both searching, filtering and paging

### LINKED ISSUE #24662: Incomplete issue search results in repository with many issues
### Description

There multiple ways to reproduce this, but one way:
* Create 50 issues with same title
* Create 1 pull request with the same title as the issues
* Searching for the title will return either 0 results in pull requests search, or only 49 results in issue search

Another way:
* Create 60 issues with same title
* Apply one label to 30 of them, and another label to the other 30
* Search for the title, filter by one of the labels, and it will return less than 30 results

The reason behind this is that that indexers will index the title, contents and comments and return up to 50 search results based on that. Filtering by issue or PR, open or closed, labels, author, .. happens afterwards. Note that pagination as in #22704 does not solve this problem.

The solution could be to make all indexers filter by and index all these issue fields. That would require adding quite a bit of code to all indexers though, every filtering option would need to be implemented in every indexer.

Alternative solutions with worse performance would be to get an unlimited number of results from the indexers, or compute a list of filter matching issue IDs to give to the indexers.

### Gitea Version

60e7963 (main)

### Can you reproduce the bug on the Gitea demo site?

Yes

### Log Gist

_No response_

### Screenshots

_No response_

### Git Version

_No response_

### Operating System

_No response_

### How are you running Gitea?

Own build, using meilisearch. But it should happen anywhere, with any indexer.

### Database

None

### PR DESCRIPTION
**Pull request by wolfogre**
Fix #24662.

Replace #24822 and #25708 (although it has been merged)

## Background

In the past, Gitea supported issue searching with a keyword and conditions in a less efficient way. It worked by searching for issues with the keyword and obtaining limited IDs (as it is heavy to get all) on the indexer (bleve/elasticsearch/meilisearch), and then querying with conditions on the database to find a subset of the found IDs. This is why the results could be incomplete.

To solve this issue, we need to store all fields that could be used as conditions in the indexer and support both keyword and additional conditions when searching with the indexer.

## Major changes

- Redefine `IndexerData` to include all fields that could be used as filter conditions.
- Refactor `Search(ctx context.Context, kw string, repoIDs []int64, limit, start int, state string)` to `Search(ctx context.Context, options *SearchOptions)`, so it supports more conditions now.
- Change the data type stored in `issueIndexerQueue`. Use `IndexerMetadata` instead of `IndexerData` in case the data has been updated while it is in the queue. This also reduces the storage size of the queue.
- Enhance searching with Bleve/Elasticsearch/Meilisearch, make them fully support `SearchOptions`. Also, update the data versions.
- Keep most logic of database indexer, but remove `issues.SearchIssueIDsByKeyword` in `models` to avoid confusion where is the entry point to search issues.
- Start a Meilisearch instance to test it in unit tests.
- Add unit tests with almost full coverage to test Bleve/Elasticsearch/Meilisearch indexer.

### CODE REVIEW

**Comment by wxiaoguang** [comment]
I think it's safer to have a separate PR for "fix: LoadAttributes context", to avoid potential conflicts (because other PRs may also need such change)

**Comment by lunny** [comment]
> I think it's safer to have a separate PR for "fix: LoadAttributes context", to avoid potential conflicts (because other PRs may also need such change)

ref #26055

**Comment by lunny** [comment]
#26128 is ready to review.

File: modules/indexer/issues/indexer.go
Comment (wxiaoguang) [inline modules/indexer/issues/indexer.go:41]: I think we can introduce a "Version" field here.

If Version=0, then use old rules.

If Version=1, then use new rules. New code will always emit Version=1 items.

**Review by wxiaoguang — APPROVED** [review]
Overall LGTM.

The `NoConditionID and sometimes MilestoneIDs = []int64{0} and sometimes MilestoneIDs = []int64{NoConditionID}` seems difficult to understand (I have seen it many times), I suspect some of them is not right?

**Review by lunny — APPROVED** [review]

**Comment by wxiaoguang** [comment]
@lunny I think it should wait for the author to take a look at these review comments first, before the unnecessary "urgent" wait-merge.

**Review by techknowlogick — DISMISSED** [review]
blocking so I get a chance to review before it gets merged.

File: modules/indexer/issues/indexer.go
Comment (wolfogre) [reply modules/indexer/issues/indexer.go:41]: Since the new struct is compatible with the old one, we could do that if we have to break it one day.

Although it looks weird, I think it's good enough.

**Comment by wolfogre** [comment]
> The `NoConditionID and sometimes MilestoneIDs = []int64{0} and sometimes MilestoneIDs = []int64{NoConditionID}` seems difficult to understand (I have seen it many times), I suspect some of them is not right?

`MilestoneIDs = []int64{0}` is for indexer searching, and `MilestoneIDs = []int64{db.NoConditionID}` is for db querying.

I don't want use `db.NoConditionID` in indexer, just keep it simple for indexer, `0` means `= 0`, `-1` means `= -1`, and a flag `NoLabelOnly`(not 0 nor -1) means `len(lables) == 0`

**Comment by wxiaoguang** [comment]
> `MilestoneIDs = []int64{0}` is for indexer searching, and `MilestoneIDs = []int64{db.NoConditionID}` is for db querying.
> 
> I don't want use `db.NoConditionID` in indexer, just keep it simple for indexer, `0` means `= 0`, `-1` means `= -1`, and a flag `NoLabelOnly`(not `0` nor `-1`) means `len(lables) == 0`
> 
> I think I'm careful in handling them (I hope I'm careful enough 🥲).

It looks pretty weird to me, but since it works, no block from my side.

**Comment by wxiaoguang** [comment]
> blocking so I get a chance to review before it gets merged.

Oops, sorry I dismissed your review (I thought it was for mine .......), I will add a "blocked" manually.

**Review by techknowlogick — APPROVED** [review]
ty :)

**Pull request merged**
